Firestore/Source/Local/FSTLocalStore.mm

```diff
@@ -156,12 +159,53 @@ - (FSTMaybeDocumentDictionary *)userDidChange:(const User &)user {
 }
 
 - (FSTLocalWriteResult *)locallyWriteMutations:(NSArray<FSTMutation *> *)mutations {
+  FIRTimestamp *localWriteTime = [FIRTimestamp timestamp];
+  DocumentKeySet keys;
+  for (FSTMutation *mutation in mutations) {
+    keys = keys.insert(mutation.key);
+  }
+
   return self.persistence.run("Locally write mutations", [&]() -> FSTLocalWriteResult * {
-    FIRTimestamp *localWriteTime = [FIRTimestamp timestamp];
-    FSTMutationBatch *batch =
-        [self.mutationQueue addMutationBatchWithWriteTime:localWriteTime mutations:mutations];
-    DocumentKeySet keys = [batch keys];
-    FSTMaybeDocumentDictionary *changedDocuments = [self.localDocuments documentsForKeys:keys];
+    // Load and apply all existing mutations. This lets us compute the current base state for
+    // all non-idempotent transforms before applying any additional user-provided writes.
+    FSTMaybeDocumentDictionary *existingDocuments = [self.localDocuments documentsForKeys:keys];
+
+    // For non-idempotent mutations (such as `FieldValue.increment()`), we record the base
+    // state in a separate patch mutation. This is later used to guarantee consistent values
+    // and prevents flicker even if the backend sends us an update that already includes our
+    // transform.
+    NSMutableArray<FSTMutation *> *baseMutations = [NSMutableArray array];
+    for (FSTMutation *mutation in mutations) {
+      FSTMaybeDocument *maybeDocument = [existingDocuments objectForKey:mutation.key];
+      if (mutation.idempotent) {
+        continue;
+      }
+
+      // Theoretically, we should only include non-idempotent fields in this field mask as this mask
+      // is used to prevent flicker for non-idempotent transforms by providing consistent base
+      // values. By including the fields for all DocumentTransforms, we incorrectly prevent rebasing
+      // of idempotent transforms (such as `arrayUnion()`) when any non-idempotent transforms are
+      // present.
+      const FieldMask *fieldMask = [mutation fieldMask];
+      if (fieldMask) {
+        FSTObjectValue *baseValues =
+            [maybeDocument isKindOfClass:[FSTDocument class]]
+                ? [((FSTDocument *)maybeDocument).data objectByApplyingFieldMask:*fieldMask]
+                : [FSTObjectValue objectValue];
+        // NOTE: The base state should only be applied if there's some existing document to
+        // override, so use a Precondition of exists=true
+        [baseMutations addObject:[[FSTPatchMutation alloc] initWithKey:mutation.key
+                                                             fieldMask:FieldMask(*fieldMask)
```

Comment:
Would it make sense to std::move the mask into the copy constructor?

Reply:
That question seems somewhat nonsensical.

If you're copying you shouldn't move since the whole point of copying is to preserve the existing value.

If you want to move don't copy.

Reply:
AFAICS, in this case `std::move` won't have any effect, because `fieldMask` points to a `const` object.

`std::move` is essentially a cast; simplifying a little bit, it's pretty much the same as `static_cast<T&&>(foo)`. The only thing `move` does is affect the overload resolution:
```cpp
void Foo(const Bar& bar); // Takes const lvalue reference
void Foo(Bar&& bar); // Takes rvalue reference
Bar bar;
Foo(bar); // Chooses the first overload (const lvalue reference)
Foo(std::move(bar)); // Chooses the second overload (rvalue reference)
```
So in this case, `move` could affect whether it's the copy constructor or the move constructor that gets selected. However, moving a `const` value results in a somewhat weird type `const Foo&&` (const rvalue reference), which cannot bind to a non-const rvalue argument; so when overload is being selected for `const Foo&&`, `Foo&&` is out of question, and `const Foo&` ends up being selected. This means that the copy constructor will be used in this case, which would have happened anyway without the move.

Note that moving an object is a commitment to never use the object again except to reassign it; moving pretty much gives up possession of the object.

Reply:
Thanks for the thorough explanation. I was worried it would create an extra copy of the FieldMask before it calls the copy-constructor, but now that I understand this better, I am no longer worried. I appreciate your explanation!